src/parser/antlr_input.cpp

```diff
@@ -501,9 +507,14 @@ std::string parseErrorHelper(const char* lineStart, int charPositionInLine, cons
 
 void AntlrInput::parseError(const std::string& message, bool eofException)
 {
-  string updatedMessage = parseErrorHelper((const char*)d_antlr3InputStream->getLineBuf(d_antlr3InputStream),
-                                           d_lexer->getCharPositionInLine(d_lexer),
-                                           message);
+  auto lineLength = d_antlr3InputStream->size(d_antlr3InputStream)
+                    - (static_cast<char*>(d_antlr3InputStream->currentLine)
+                       - static_cast<char*>(d_antlr3InputStream->data));
```

Comment:
Hm, so this may actually the issue that I've run into previously when trying to fix the issue: doesn't this assume that all the lines are allocated linearly in the memory?

Reply:
You mean ANTLR could use the buffer in a cyclic fashion so that the current line is followed by something else than the succeeding line, right?
I don't think this would be a problem, but may highlight a naming issue.
Note that printing still stops when we hit a `\n` and `lineLength` is only used when there is no `\n` between `lineStart` (or rather `getLineBuf()` and the end of the buffer. A name like `remainingBufferSize` (instead of `lineLength`) may be more appropriate here...

Reply:
Hm, the problem is that the individual lines can be at completely arbitrary places in memory, when using the buffered input. It may be the case that your solution works when reading from a file but fails when `cat`ing it through `stdin`. Could you try that?

For buffered inputs, the lines are stored [here](https://github.com/CVC4/CVC4/blob/8374a8dd4bf740bf26748c1dbe1616ad798cf624/src/parser/line_buffer.h#L67) and allocated [here](https://github.com/CVC4/CVC4/blob/8374a8dd4bf740bf26748c1dbe1616ad798cf624/src/parser/line_buffer.cpp#L82-L86). I believe that the `LineBuffer` is only used when reading from `stdin` (see [this](https://github.com/CVC4/CVC4/blob/8374a8dd4bf740bf26748c1dbe1616ad798cf624/src/main/driver_unified.cpp#L425)).

Reply:
Ah, good point. It seems we can use `sizeBuf` instead of `size()` though:

> $ cat test_all.sh 
> #!/bin/bash
> bin/cvc4 test_4866.smt2
> bin/cvc4 --lang=smt2 < test_4866.smt2
> bin/cvc4 --incremental test_4069.smt2
> bin/cvc4 --incremental --lang=smt2 < test_4069.smt2

> $ ./test_all.sh 
> (error "Parse Error: test_4866.smt2:3.6: expected SMT-LIBv2 command, got `reset'.
> 
>   (reset)
>    ^
> ")
> (error "Parse Error: <stdin>:2.6: expected SMT-LIBv2 command, got `reset'.
> 
>   (reset)
>    ^
> ")
> (error "Parse Error: test_4069.smt2:323.646: Unexpected token: '<<EOF>>'.")
> (error "Parse Error: <stdin>:324.0: Unexpected token: '<<EOF>>'.")

(`bin/cvc4` being compiled with `Debug --asan --ubsan`)

Reply:
Hmm, I am actually mostly worried about `static_cast<char*>(d_antlr3InputStream->currentLine) - static_cast<char*>(d_antlr3InputStream->data)` because I am worried that that is not accurate when we allocate the lines in a way that is not adjacent. This may only happen when you do it truly interactively, i.e. you only feed a command to CVC4 after you get an answer for a previous one (if you pass the whole file to `stdin` it may be the case that we allocate all the lines at once and you don't have the problem, I'd need to check).

Reply:
I agree that this solution may not be completely accurate if the lines are not in order (within the buffer). However, this also requires that the line we are looking at is not properly terminated, and it is directly followed (i.e. not separated by a newline character) by another line.
But even in this case, this PR improves the current situation: instead of memory errors we may only print another line after the one we are actually interested in.
I'd be glad if you can come up with an even better solution, but this `ANTLR3_INPUT_STREAM_struct` really does not give you a lot to work with...